diff --git a/build/order_by_dep.awk b/build/order_by_dep.awk
index 1e71ea2069..3da32d8830 100644
--- a/build/order_by_dep.awk
+++ b/build/order_by_dep.awk
@@ -37,6 +37,11 @@ function get_module_index(name,  i)
 function do_deps(mod_idx,        module_name, mod_name_len, dep, ext, val, depidx)
 {
 	module_name = mods[mod_idx];
+	# TODO: real skip zend extension
+	if (module_name == "opcache") {
+		delete mods[mod_idx];
+		return;
+	}
 	mod_name_len = length(module_name);
 
 	for (ext in mod_deps) {
diff --git a/ext/opcache/ZendAccelerator.c b/ext/opcache/ZendAccelerator.c
index 5f6b854d47..ea15c0d5bc 100644
--- a/ext/opcache/ZendAccelerator.c
+++ b/ext/opcache/ZendAccelerator.c
@@ -91,7 +91,10 @@ typedef int gid_t;
 #include <immintrin.h>
 #endif
 
+#ifdef COMPILE_DL_OPCACHE
+// avoid symbol conflict
 ZEND_EXTENSION();
+#endif
 
 #ifndef ZTS
 zend_accel_globals accel_globals;
@@ -4808,7 +4811,11 @@ static int accel_finish_startup(void)
 	return SUCCESS;
 }
 
+#ifdef COMPILE_DL_OPCACHE
 ZEND_EXT_API zend_extension zend_extension_entry = {
+#else
+zend_extension opcache_zend_extension_entry = {
+#endif
 	ACCELERATOR_PRODUCT_NAME,               /* name */
 	PHP_VERSION,							/* version */
 	"Zend Technologies",					/* author */
diff --git a/ext/opcache/config.m4 b/ext/opcache/config.m4
index 2a83fa2455..69274ae1a5 100644
--- a/ext/opcache/config.m4
+++ b/ext/opcache/config.m4
@@ -21,7 +21,8 @@ PHP_ARG_ENABLE([opcache-jit],
 if test "$PHP_OPCACHE" != "no"; then
 
   dnl Always build as shared extension
-  ext_shared=yes
+  dnl why?
+  dnl ext_shared=yes
 
   if test "$PHP_HUGE_CODE_PAGES" = "yes"; then
     AC_DEFINE(HAVE_HUGE_CODE_PAGES, 1, [Define to enable copying PHP CODE pages into HUGE PAGES (experimental)])
@@ -327,7 +328,7 @@ int main() {
 	shared_alloc_mmap.c \
 	shared_alloc_posix.c \
 	$ZEND_JIT_SRC,
-	shared,,"-Wno-implicit-fallthrough -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1",,yes)
+  $ext_shared,,"-Wno-implicit-fallthrough -DZEND_ENABLE_STATIC_TSRMLS_CACHE=1",,yes)
 
   PHP_ADD_EXTENSION_DEP(opcache, pcre)
 
diff --git a/main/main.c b/main/main.c
index 8c16f01b11..f04b86a46e 100644
--- a/main/main.c
+++ b/main/main.c
@@ -2011,6 +2011,18 @@ void dummy_invalid_parameter_handler(
 }
 #endif
 
+// this can be moved to other place
+#ifndef COMPILE_DL_OPCACHE
+extern zend_extension opcache_zend_extension_entry;
+extern void zend_register_extension(zend_extension *new_extension, void *handle);
+
+int zend_load_static_extensions(void)
+{
+	zend_register_extension(&opcache_zend_extension_entry, NULL /*opcache cannot be unloaded*/);
+	return 0;
+}
+#endif
+
 /* {{{ php_module_startup */
 int php_module_startup(sapi_module_struct *sf, zend_module_entry *additional_modules, uint32_t num_additional_modules)
 {
@@ -2253,6 +2265,9 @@ int php_module_startup(sapi_module_struct *sf, zend_module_entry *additional_mod
 	   ahead of all other internals
 	 */
 	php_ini_register_extensions();
+#ifndef COMPILE_DL_OPCACHE
+	zend_load_static_extensions();
+#endif
 	zend_startup_modules();
 
 	/* start Zend extensions */
